_posts/2015-09-28-service-discovery.md

```diff
@@ -0,0 +1,57 @@
+
+#How does it work?
+A service discovery solution usually consists of three core building blocks:
+
+* A *service registry* which holds the data about currently available services and instances
+* A *health-check* mechanism to monitor service health
+* A *lookup* or *routing* mechanism to connect to services
+
+## Service Registry
+At the heart of the service discovery infrastructure, there is a so-called service registry. This registry has the knowledge about all available services and their instances. It gains this knowledge by providing an API to register and deregister hosts or service instances, so the services have a central point to make themselves available to the public.
+
+####Consul
+We chose [Consul](https://consul.io/) as our service registry implementation. Consul provides the core functionality of a service registry by utilizing an agent-based setup. This means, an agent, which is kind of a daemon process, runs on every machine that provides services. Here is a short abstract of Consul's introduction documentation to get to know the most important parts:
+
+>Consul is a distributed, highly available system. [...]
+>Every node that provides services to Consul runs a *Consul agent*.  Running an agent is not required for discovering other services or getting/setting key/value data. The agent is responsible for health checking the services on the node as well as the node itself.
+>The agents talk to one or more *Consul servers*. The Consul servers are where data is stored and replicated. The servers themselves elect a leader.
+>[...]
+>Components of your infrastructure that need to discover other services or nodes can query any of the Consul servers
+>or any of the Consul agents. The agents forward queries to the servers automatically.
+
+For more details, see the full [introduction](https://www.consul.io/intro/index.html)  and the [in-depth architecture overview](https://www.consul.io/docs/internals/architecture.html) on the Consul web site.
+
+##Health-Checks
+The data inside the service registry only provides real value if it is up to date. Just knowing there once was a service available at a certain location is not enough. Therefore, checking the availability or health status of all service instances is crucial for service discovery.
+
+In our setup, the health checking is also done by Consul. The servers check the general availability of the agents, and therefore the hosts. In addition, you can register different types of service checks with an agent, which then perform these checks against the services registered locally in fixed intervals. Whenever such a check fails, the according service or host will be marked as unavailable, and therefore the service registry will no longer promote the faulty instance any longer.
```

Comment:
... service checks with an agent, which then perform these checks 
What's the relation here? Does the verb "perform" relate to "service checks" or "an agent"? Singular vs Plural